language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,410 @@
+    mode_type: ModeType,
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    about = "Replay transactions using state prefetching",
+    rename_all = "kebab-case"
+)]
+pub struct PrefetchedReplayCLI {
+    #[structopt(long, default_value = DEFAULT_NODE_URL)]
+    rpc_url: String,
+
+    #[structopt(subcommand)]
+    cmd: Command,
+}
+
+#[derive(Debug, StructOpt)]
+pub enum Command {
+    /// Replay transactions starting from version `start` to `start + limit`.
+    #[structopt(name = "replay-transactions")]
+    ReplayTransactions { start: Version, limit: u64 },
+
+    /// Replay `account` transactions starting from version `start` to `start + limit`.
+    #[structopt(name = "replay-account-transactions")]
+    ReplayAccountTransactions {
+        account: AccountAddress,
+        start: Version,
+        limit: u64,
+    },
+
+    /// Execute serialized transactions from `txs_path`
+    #[structopt(name = "execute-serialized-transactions")]
+    ExecuteSerializedTransactions { txs_path: PathBuf },
+}
+
+pub struct PrefetchTransactionReplayer {
+    config: PrefetchTransactionReplayerConfig,
+    validator: Box<dyn DiemValidatorInterface>,
+    debugger: DiemDebugger,
+    rpc_client: BlockingClient,
+}
+
+impl StateView for OnDiskStateViewReader {
+    fn get(&self, ap: &AccessPath) -> Result<Option<Vec<u8>>> {
+        match ap.get_path() {
+            access_path::Path::Code(mid) => self.view.get_module_bytes(&mid),
+            access_path::Path::Resource(tag) => self.view.get_resource_bytes(ap.address, tag),
+        }
+    }
+
+    fn is_genesis(&self) -> bool {
+        false
+    }
+}
+
+fn ensure_dir_does_not_exist(dir: &Path) -> Result<()> {
+    if dir.exists() {
+        Err(format_err!("{:?} must be empty as it will be created", dir))
+    } else {
+        Ok(())
+    }
+}
+
+impl PrefetchTransactionReplayer {
+    pub fn from_config(config: PrefetchTransactionReplayerConfig) -> Result<Self> {
+        ensure_dir_does_not_exist(&config.build_dir)?;
+        ensure_dir_does_not_exist(&config.storage_dir)?;
+
+        let debugger = DiemDebugger::json_rpc_with_config(
+            &config.rpc_url,
+            config.build_dir.clone(),
+            config.storage_dir.clone(),
+        )?;
+
+        let rpc_client = BlockingClient::new(&config.rpc_url);
+
+        // NOTE: we could probably avoid creating a second validator
+        // by making the one in `DiemDebugger` accessible publicly
+        let validator = Box::new(JsonRpcDebuggerInterface::new(&config.rpc_url)?);
+        Ok(PrefetchTransactionReplayer {
+            config,
+            validator,
+            debugger,
+            rpc_client,
+        })
+    }
+
+    fn get_keys_read_for_tx(
+        &self,
+        live_state: &impl MoveResolver,
+        tx: &Transaction,
+        rw_analysis: &ReadWriteSetAnalysis,
+    ) -> Result<Vec<ResourceKey>> {
+        let signed_tx = match tx {
+            Transaction::UserTransaction(u) => u,
+            t => {
```

Comment:
I think we can still handle block prologue transaction here but that may not be that useful.